cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -323,9 +328,42 @@ private MetadataEntity getMetadataEntityFromPath(String uri, @Nullable  String e
       }
       curIndex += 2;
     }
+    builder = makeBackwardCompatible(builder);
     return builder.build();
   }
 
+  /**
+   * If the MetadataEntity Builder key-value represent an application or artifact which is versioned in CDAP i.e. their
+   * metadata entity representation ends with 'version' rather than 'application' or 'artifact' and the type is also
+   * set to 'version' then for backward compatibility of rest end points return an updated builder which has the
+   * correct type. This is only needed in 5.0 for backward compatibility of the rest endpoint.
+   * From 5.0 and later the rest end point must be called with a query parameter which specify the type of the
+   * metadata entity if the type is not the last key. (CDAP-13678)
+   */
+  @VisibleForTesting
+  static MetadataEntity.Builder makeBackwardCompatible(MetadataEntity.Builder builder) {
+    MetadataEntity entity = builder.build();
+    Iterator<MetadataEntity.KeyValue> iterator = entity.iterator();
+    List<MetadataEntity.KeyValue> entityKeyValues = StreamSupport.stream(
```

Comment:
simpler to do `StreamSupport.stream(entity.spliterator(), false).collect(Collectors.toList())`